bzerolib/connection/websocket/websocket.go

```
@@ -0,0 +1,126 @@
+package websocket
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"bastionzero.com/bctl/v1/bzerolib/logger"
+	gorilla "github.com/gorilla/websocket"
+)
+
+var websocketUrlScheme = "wss"
+
+type IWebsocket interface {
+	Done() <-chan struct{}
+	Inbound() <-chan *[]byte
+	Dial(websocketUrl *url.URL, ctx context.Context) (err error)
+	Send(message []byte) error
+	Close()
+}
+
+type Websocket struct {
+	logger   *logger.Logger
+	doneChan chan struct{}
+	isDead   bool
+
+	client *gorilla.Conn
+
+	// Received messages
+	inbound chan *[]byte
+
+	// Messages to be sent
+	outbound chan *[]byte
+}
+
+func New(logger *logger.Logger) *Websocket {
+	return &Websocket{
+		doneChan: make(chan struct{}),
+		logger:   logger,
+		inbound:  make(chan *[]byte, 200),
+		outbound: make(chan *[]byte, 200),
+	}
+}
+
+func (w *Websocket) Close() {
+	if !w.isDead {
+		w.isDead = true
+		close(w.doneChan)
+	}
+}
+
+func (w *Websocket) Done() <-chan struct{} {
+	return w.doneChan
+}
+
+func (w *Websocket) Inbound() <-chan *[]byte {
+	return w.inbound
+}
+
+func (w *Websocket) Send(message []byte) error {
+	if err := w.client.WriteMessage(gorilla.TextMessage, message); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *Websocket) Dial(websocketUrl *url.URL, ctx context.Context) (err error) {
+	// Reinitialize our variables every time in case this is post death
+	if w.isDead {
+		w.doneChan = make(chan struct{})
+		w.isDead = false
+	}
+
+	// Make sure url scheme is correct
+	websocketUrl.Scheme = websocketUrlScheme
```

Would it be better to return an error if given an invalid url scheme?

---

I put some constants above to make for tighter controls around it but I feel it might be unecessary to check for an invalid scheme here. Let me know if you still think we should check here

---

63ef203f774de99cd6320952697b368303f1d64b